packages/shylu/shylu_node/tacho/unit-test/Tacho_TestSolver.cpp

```diff
@@ -0,0 +1,242 @@
+#include "Tacho_CrsMatrixBase.hpp"
+#include "Tacho_Driver.hpp"
+#include "Tacho_MatrixMarket.hpp"
+
+
+/// matrix generator
+template <typename CrsMatrixBaseTypeHost>
+int generate2dLaplace(int nx, CrsMatrixBaseTypeHost &A) {
+  // generate Laplace matrix on 5pt stencile
+  int n = nx*nx;
+  int nnz = n + 4*nx*(nx-1);
+
+  using value_type = typename CrsMatrixBaseTypeHost::value_type;
+  typename CrsMatrixBaseTypeHost::size_type_array    ap("ap", n+1);
+  typename CrsMatrixBaseTypeHost::ordinal_type_array aj("aj", nnz);
+  typename CrsMatrixBaseTypeHost::value_type_array   ax("ax", nnz);
+
+  nnz = 0;
+  ap(0) = 0;
+  for (int i=0; i<nx; i++) {
+    for (int j=0; j<nx; j++) {
+      int k = i*nx + j;
+      if (i > 0) {
+        aj(nnz) = k-nx;
+        ax(nnz) = value_type(1.0);
+        nnz ++;
+      }
+      if (j > 0) {
+        aj(nnz) = k-1;
+        ax(nnz) = value_type(1.0);
+        nnz ++;
+      }
+      aj(nnz) = k;
+      ax(nnz) = value_type(4.0);
+      nnz++;
+      if (j < nx-1) {
+        aj(nnz) = k+1;
+        ax(nnz) = value_type(1.0);
+        nnz ++;
+      }
+      if (i < nx-1) {
+        aj(nnz) = k+nx;
+        ax(nnz) = value_type(1.0);
+        nnz ++;
+      }
+      ap(k+1) = nnz;
+    }
+  }
+  A.clear();
+  A.setExternalMatrix(n, n, nnz, ap, aj, ax);
+  return nnz;
+}
+
+
+// main test driver
+template <typename value_type>
+int driver(std::string file, std::string method_name, int variant, int nrhs) {
+  int nx = 10;
+  int method = 1; // 1 - Chol, 2 - LDL, 3 - SymLU
+  if (method_name == "chol")
+    method = 1;
+  else if (method_name == "ldl")
+    method = 2;
+  else if (method_name == "lu")
+    method = 3;
+  else {
+    std::cout << "Error: not supported solution method\n";
+    return -1;
+  }
+
+  using device_type = typename Tacho::UseThisDevice<Kokkos::DefaultExecutionSpace>::type;
+  using host_device_type = typename Tacho::UseThisDevice<Kokkos::DefaultHostExecutionSpace>::type;
+
+  const bool detail = false;
+  const bool verbose = false;
+  Tacho::printExecSpaceConfiguration<typename device_type::execution_space>("DeviceSpace", detail);
+  Tacho::printExecSpaceConfiguration<typename host_device_type::execution_space>("HostSpace", detail);
+  std::cout << std::endl << "    --------------------- " << std::endl;
+  std::cout << "     Method Name:: " << method_name << std::endl;
+  std::cout << "     Solver Type:: " << variant << std::endl;
+  std::cout << "          # RHSs:: " << nrhs << std::endl;
+  std::cout << "          Matrix:: " << file;
+  std::cout << std::endl << "    --------------------- " << std::endl << std::endl;
+
+  int r_val = 0;
+  try {
+    /// crs matrix format and dense multi vector
+    using CrsMatrixBaseTypeHost = Tacho::CrsMatrixBase<value_type, host_device_type>;
+    using DenseMultiVectorType = Kokkos::View<value_type **, Kokkos::LayoutLeft, device_type>;
+
+    /// read a spd matrix of matrix market format
+    CrsMatrixBaseTypeHost A;
+    if (file != "") {
```

Comment:
Alternative (minor suggestion for readability, but might not be the preference for this check)
```suggestion
    if (!file.empty()) {
```